File: pcr2/src/main/java/edu/monash/humanise/smartcity/pcr2/PayloadService.java
package edu.monash.humanise.smartcity.pcr2;

import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@AllArgsConstructor
public class PayloadService {
    private final PayloadRepository payloadRepository;
    public void decodeUplinkPayload(PayloadUplinkRequest payloadRequest) {
        try{
            Decoder decoder = new Decoder(payloadRequest.data());
            decoder.decode();

            Payload payload = Payload.builder()
                    .deviceName(payloadRequest.deviceName())
                    .data(payloadRequest.data())
                    .devEUI(payloadRequest.devEUI())
                    .ltr(decoder.getLtr())
                    .rtl(decoder.getRtl())
                    .cpuTemp(decoder.getCpuTemp())
                    .build();
            payloadRepository.save(payload);

        } catch (Exception e) {
            log.error("Got exception in service. " + e.getMessage());
        }
    }
}

File: pcr2/src/main/kotlin/edu/monash/humanise/smartcity/pcr2/PayloadService.kt
package edu.monash.humanise.smartcity.pcr2

import org.springframework.stereotype.Service
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset

@Service
class PayloadService(private val payloadRepository: PayloadRepository) {
    fun decodeUplinkPayload(payloadRequest: PayloadUplinkRequest) {
        val timestamp =
            OffsetDateTime.ofInstant(Instant.ofEpochMilli(payloadRequest.timestampMilliseconds), ZoneOffset.UTC)
        val decoded = Decoder.decode(payloadRequest.data)
        val payload = Payload(
            payloadRequest.deviceName,
            payloadRequest.data,
            payloadRequest.devEUI,
            timestamp,
            decoded.ltr,
            decoded.rtl,
            decoded.cpuTemp
        )
        payloadRepository.save(payload)
    }
}
